### PR TITLE
NC-603

### PR DESCRIPTION
**Pull request by Joeper214**
ref: https://skyarch.backlog.jp/view/NC-603

### Goal 
- [x] check project count on clients page and remove delete button if it has 1 or many projects
- [x] check infrastructure count on prject page and remove delete button if it has 1 or many infrastructes
- [x] Add indicator of number of projects in client page
- [x] Add indicator of number of infrastructures in project page

- [x] iT18N
- [x] Write tests

### CODE REVIEW

File: app/assets/javascripts/projects.js [hunk 1]
```diff
@@ -1,4 +1,4 @@
-//
+\//
```
Comment (skyarch-networks) [inline app/assets/javascripts/projects.js]: __[jshint]__
*  Unexpected '\'.

**Comment by Joeper214** [comment]
Please Review and Merge PR. @pocke

**Comment by pocke** [comment]
@Joeper214 please resolve conflict :bow:

**Comment by pocke** [comment]
Thanks for fixing conflict :bow:

File: app/assets/javascripts/clients.js [hunk 2]
```diff
@@ -113,7 +114,7 @@
     el: '#indexElement',
     data: {
       searchQuery: '',
-      gridColumns: ['code','name', 'id'],
+      gridColumns: ['code','name', 'nums','id'],
```
Comment (pocke) [inline app/assets/javascripts/clients.js]: :+1:

File: app/helpers/clients_helper.rb [hunk 3]
```diff
@@ -18,6 +19,8 @@ def button_edit_client(client)
 
   def button_delete_client(client)
     return nil unless Pundit.policy(current_user, client).destroy?
+    return nil if client.projects.count > 0
```
Comment (pocke) [inline app/helpers/clients_helper.rb]: ( here and projects_helper )

I think `disabled` is better than hiding.
User can't delete the client always if user doesn't have any permissions.
But when `projects.count` isn't 0, user can delete the client if user delete projects.

**Comment by pocke** [comment]
In controller and model, please add validation for number of projects/infras.



- [ActiveRecord::Associations::ClassMethods](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many-label-Options)
- [Rails - dependent: :restrict_with_error と :restrict_with_exception の違い - Qiita](http://qiita.com/jnchito/items/3456ce734ef41d216ecd) (Japanese)

File: app/helpers/clients_helper.rb (hunk 3 above)
Comment (Joeper214) [inline app/helpers/clients_helper.rb]: Yes. but did you mean that if `projects == 0`. client can delete projects. while if `projects > 0`, they cannot delete the client?

Comment (Joeper214) [inline app/helpers/clients_helper.rb]: because this code `return nil if client.projects.count > 0` checks if the client has one or many projects.

Comment (pocke) [inline app/helpers/clients_helper.rb]: Yes.

Comment (Joeper214) [inline app/helpers/clients_helper.rb]: :ok_hand:

**Comment by Joeper214** [comment]
Is the purpose if validation is when deleting a project/client?

**Comment by pocke** [comment]
yes, we should change this line 

https://github.com/skyarch-networks/skyhopper/blob/master/app/models/client.rb#L10

**Comment by Joeper214** [comment]
:+1:   maybe the line will be. `has_many :projects, dependent: :restrict_with_error`

**Comment by pocke** [comment]
It's ok :100: 
but I think we should also change tests

**Comment by Joeper214** [comment]
yes. I'll check it out.

**Comment by Joeper214** [comment]
But I still don't know how to write test. hmmmm..

**Comment by pocke** [comment]
Thanks.

```ruby
project = create :project
infra = create :infra
project.infrastructures = [infra]
project.destroy # => assert error is an ActiveRecord::DeleteRestrictionError

infra.destroy
project.destroy # => assert error is nil
```

We should test the above code.

https://www.relishapp.com/rspec/rspec-expectations/v/2-11/docs/built-in-matchers/raise-error-matcher
Maybe this URL is helpful

**Comment by Joeper214** [comment]
it's in the 
 - `project_controller_spec.rb`
 - `client_controller_spec.rb`

right?

**Comment by pocke** [comment]
Hmm... no. 

- `app/models/project.rb`
- `app/models/client.rb`

**Comment by Joeper214** [comment]
:+1:

**Comment by pocke** [comment]
:tada:

**Comment by Joeper214** [comment]
:tada:  ありがとう！

**Pull request merged**